File: src/08_Calculator/CalculatorServer.java
import java.io.*;
import java.net.*;
import java.util.*;

public class CalculatorServer {
    public static void main(String[] args) {
        ServerSocket serverSocket = null;
        Socket socket = null;
        BufferedReader input = null;
        PrintWriter output = null;

        try {
          
            serverSocket = new ServerSocket(5000);
            System.out.println("Server is running...");

            socket = serverSocket.accept();
            System.out.println("Client connected.");

            
            input = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            output = new PrintWriter(socket.getOutputStream(), true);

            
            String expression;
            while ((expression = input.readLine()) != null) {
                System.out.println("Received expression: " + expression);

            
                String result = evaluateExpression(expression);
                output.println(result);
            }
        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            try {
                if (input != null) input.close();
                if (output != null) output.close();
                if (socket != null) socket.close();
                if (serverSocket != null) serverSocket.close();
            } catch (IOException e) {
                System.out.println("Error closing resources: " + e.getMessage());
            }
        }
    }

    // Method to evaluate simple mathematical expressions
    private static String evaluateExpression(String expression) {
        try {
            // Remove all spaces from the expression
            expression = expression.replaceAll("\\s+", "");

            // Use Java's built-in scripting engine to evaluate basic expressions (e.g., +, -, *, /)
            double result = evaluateBasicExpression(expression);
            return String.valueOf(result);
        } catch (Exception e) {
            return "Invalid Expression";
        }
    }

    // Basic expression evaluator
    private static double evaluateBasicExpression(String expression) throws Exception {
        // Stack to store numbers and operators
        Stack<Double> numbers = new Stack<>();
        Stack<Character> operators = new Stack<>();

        int n = expression.length();
        for (int i = 0; i < n; i++) {
            char ch = expression.charAt(i);

            // If the character is a number, parse it
            if (Character.isDigit(ch)) {
                StringBuilder sb = new StringBuilder();
                // Handle multi-digit numbers
                while (i < n && (Character.isDigit(expression.charAt(i)) || expression.charAt(i) == '.')) {
                    sb.append(expression.charAt(i++));
                }
                i--; // Go back by one position
                numbers.push(Double.parseDouble(sb.toString()));
            } else if (ch == '(') {
                // Push the opening bracket to the operators stack
                operators.push(ch);
            } else if (ch == ')') {
                // Solve the entire sub-expression within brackets
                while (operators.peek() != '(') {
                    numbers.push(applyOperation(operators.pop(), numbers.pop(), numbers.pop()));
                }
                operators.pop(); // Remove the '('
            } else if (ch == '+' || ch == '-' || ch == '*' || ch == '/') {
                // While the top of the operator stack has the same or greater precedence, apply the operator
                while (!operators.isEmpty() && hasPrecedence(ch, operators.peek())) {
                    numbers.push(applyOperation(operators.pop(), numbers.pop(), numbers.pop()));
                }
                operators.push(ch);
            }
        }

        // Apply the remaining operators in the stack
        while (!operators.isEmpty()) {
            numbers.push(applyOperation(operators.pop(), numbers.pop(), numbers.pop()));
        }

        // The result is the last number in the stack
        return numbers.pop();
    }

    // Check if the current operator has precedence over the top of the stack
    private static boolean hasPrecedence(char currentOp, char topOp) {
        if (topOp == '(' || topOp == ')') {
            return false;
        }
        if ((currentOp == '*' || currentOp == '/') && (topOp == '+' || topOp == '-')) {
            return false;
        }
        return true;
    }

    // Apply the operation to two operands
    private static double applyOperation(char op, double b, double a) {
        switch (op) {
            case '+':
                return a + b;
            case '-':
                return a - b;
            case '*':
                return a * b;
            case '/':
                if (b == 0) throw new UnsupportedOperationException("Cannot divide by zero");
                return a / b;
        }
        return 0;
    }
}

File: src/08_Calculator/CalculatorServer.py
import socket

def has_precedence(op1, op2):
    """Check if op2 has higher or equal precedence compared to op1."""
    if op2 in ('(', ')'):
        return False
    if (op1 in ('*', '/') and op2 in ('+', '-')):
        return False
    return True

def apply_operation(op, b, a):
    """Apply an arithmetic operation."""
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    elif op == '/':
        if b == 0:
            raise ZeroDivisionError("Cannot divide by zero")
        return a / b
    return 0

def evaluate_expression(expression):
    """Evaluate mathematical expressions safely using stacks."""
    expression = expression.replace(" ", "")
    numbers = []
    operators = []
    i = 0
    while i < len(expression):
        ch = expression[i]

        if ch.isdigit() or ch == '.':
            num_str = []
            while i < len(expression) and (expression[i].isdigit() or expression[i] == '.'):
                num_str.append(expression[i])
                i += 1
            numbers.append(float("".join(num_str)))
            continue
        elif ch == '(':
            operators.append(ch)
        elif ch == ')':
            while operators and operators[-1] != '(':
                numbers.append(apply_operation(operators.pop(), numbers.pop(), numbers.pop()))
            operators.pop()
        elif ch in "+-*/":
            while operators and has_precedence(ch, operators[-1]):
                numbers.append(apply_operation(operators.pop(), numbers.pop(), numbers.pop()))
            operators.append(ch)
        i += 1

    while operators:
        numbers.append(apply_operation(operators.pop(), numbers.pop(), numbers.pop()))

    return numbers.pop()

def start_server():
    """Start the calculator server."""
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind(("localhost", 5000))
    server_socket.listen(1)
    print("Server is running on port 5000...")

    conn, addr = server_socket.accept()
    print(f"Client connected from {addr}")

    with conn:
        while True:
            data = conn.recv(1024).decode()
            if not data:
                break
            print(f"Received expression: {data}")
            try:
                result = str(evaluate_expression(data))
            except Exception as e:
                result = f"Error: {str(e)}"
            conn.sendall(result.encode())

if __name__ == "__main__":
    start_server()
